File: algorithms/go/53-maximum-subarray/maximum-subarray.go
package problem53

// maxSubArray 动态规划
// definition: f(x) 为代表以第 i 个数结尾的连续子数组的最大和
// f(i) = max{f(i - 1) + num[i], num[i]}
func maxSubArray(nums []int) int {
	cur := nums[0]
	ans := nums[0]
	for _, num := range nums[1:] {
		if cur > 0 {
			cur = cur + num
		} else {
			cur = num
		}
		if cur > ans {
			ans = cur
		}
	}
	return ans
}

File: algorithms/53-maximum-subarray/maximum-subarray.py
"""
53. maximum subarray

最大子序和
"""

from typing import List


def max_subarray(nums: List[int]) -> int:
    """
    动态规划
    设 dp[i] 表示为以 i 下标所在元素结尾的最大子序和，那么 dp[i]=max(num[i], dp[i-1]+nums[i])
    因为我们只需要保留前一个状态，所以也无需数组。
    """
    ans = nums[0]
    prev = ans

    n = len(nums)
    for i in range(1, n):
        prev = max(prev + nums[i], nums[i])
        if prev > ans:
            ans = prev

    return ans
